reader_concurrency_semaphore.cc

```diff
@@ -1073,7 +1073,23 @@ future<> reader_concurrency_semaphore::do_wait_admission(reader_permit permit, r
         _execution_loop_future.emplace(execution_loop());
     }
 
-    const auto admit = can_admit_read(permit).decision;
+    const auto [admit, why] = can_admit_read(permit);
+    switch (why) {
+        case reason::all_ok:
+            break;
+        case reason::ready_list:
+            ++_stats.reads_queued_because_ready_list;
+            break;
+        case reason::used_permits:
+            ++_stats.reads_queued_because_used_permits;
+            break;
+        case reason::memory_resources:
+            ++_stats.reads_queued_because_memory_resources;
+            break;
+        case reason::count_resources:
+            ++_stats.reads_queued_because_count_resources;
+            break;
+    }
```

Comment:
Can we avoid using _dedicated_ conditionals to bump up stats? It may look ugly, but maybe something like

```c++
struct stats {
    uint64_t reads_queued_because[NR] = {};
};
...
++_stats.reads_queued_because[static_cast<int>(why)];
```

instead?

Or pass stats reference right into `can_admit_read` so that it could use its existing if-branches to ++ on necessary counters?

Reply:
Won't the switch be compiled to a jump table as well? No need to hand-code a jump table.

I didn't want a method which just answers some question to also bump metrics behind the scenes. Not all callers want to bump metrics. I could add a `bool bump_metrics` but that is ugly.

Reply:
Interesting https://godbolt.org/z/W5xG9of7T  Increasing the number of reasons to 6 makes clang produce ... sort of a jump table :woozy_face: but still much uglier than array-based stats updating

Reply:
Ok, I'll refactor this into a jump table myself.

Reply:
I guess both @xemul and I share an allergy to conditionals for updating stats and we both caught it in the same place.

Reply:
I should've known better than to add such obvious trigger code. Guess I like switches too much.
I will move the updating into the method. I don't want to add a `uint64 _dev_null;` member.

Reply:
Actually I cannot move the metrics bumping into the method. Not all callers want the metrics bumped, and so this would result in one if per metric bump site.

Reply:
ok